third_party/pcsc-lite/naclport/server/src/readerfactory_nacl.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright 2016 Google Inc.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+// 3. The name of the author may not be used to endorse or promote products
+//    derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+// IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file contains a replacement function for the original readerfactory.c
+// PC/SC-Lite internal implementation.
+
+#include <ppapi/cpp/instance.h>
+#include <ppapi/cpp/var.h>
+#include <ppapi/cpp/var_dictionary.h>
+
+#include <google_smart_card_common/pp_var_utils/construction.h>
+#include <google_smart_card_pcsc_lite_server/global.h>
+
+extern "C" {
+#include "readerfactory.h"
+
+LONG RFAddReaderOriginal(const char *readerNameLong, int port,
+    const char *library, const char *device);
+LONG RFRemoveReaderOriginal(const char *readerName, int port);
+}
+
+namespace {
+
+const char kTypeMessageKey[] = "type";
+const char kDataMessageKey[] = "data";
+const char kReaderInitAddMessageType[] = "reader_init_add";
+const char kReaderFinishAddMessageType[] = "reader_finish_add";
+const char kReaderRemoveMessageType[] = "reader_remove";
+const char kNameMessageKey[] = "readerName";
+const char kPortMessageKey[] = "port";
+const char kDeviceMessageKey[] = "device";
+const char kReturnCodeMessageKey[] = "returnCode";
+
+void post_message(const char* type, const pp::VarDictionary& message_data)
+{
+  pp::VarDictionary message;
+  message.Set(kTypeMessageKey, type);
+  message.Set(kDataMessageKey, message_data);
+
+  const google_smart_card::PPInstanceHolder* pp_instance_holder =
+      google_smart_card::GetGlobalPPInstanceHolder();
+  pp::Instance* const instance = pp_instance_holder->GetPPInstance();
+  instance->PostMessage(message);
```

Comment:
Looking at this code and at how RFAddReader/RFRemoveReader is used, it seems that some synchronization primitive would be required somewhere.
RFAddReader/RFRemoveReader are called on a background thread that is never finished, and it's not trivial to stop it before pp::Instance destruction begins.
So we would have to ensure that the code here works correctly with a possible concurrent pp::Instance destruction.

The similar problem was already solved in the past: see the LibusbOverChromeUsbGlobal class and how its lifetime is managed in the PpInstance class (in file smart_card_connector_app/src/pp_module.cc).
In short, the solution is:

1. Continue to have the singleton as a unique_ptr member of the PpInstance class.
2. Suppress destroying of the singleton in PpInstance destructor, so that any code that operates with it in a background thread still doesn't explode.
3. Add a method to the singleton class that "detaches" it from the pp::Instance object (essentially, the method should clear the pointer to pp::Instance in a thread-safe manner). This method should be called from the PpInstance destructor.

This solution ensures that, on one side, the code that uses the singleton won't explode during shutdown, and, on the other side, that it won't use the invalid pp::Instance* pointer.